atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventLog.java

```diff
@@ -120,6 +121,13 @@ private IdentifiedVersion getLatestVersionAndVerify(IdentifiedVersion endVersion
 
     private void processSuccess(LockWatchStateUpdate.Success success) {
         Preconditions.checkState(latestVersion.isPresent(), "Must have a known version to process successful updates");
+        Optional<IdentifiedVersion> snapshotVersion = snapshot.getSnapshotVersion();
+        Preconditions.checkState(snapshotVersion.isPresent(),
+                "Must have a snapshot before processing successful updates");
+
+        if (success.lastKnownVersion() < snapshotVersion.get().version()) {
+            throw new TransactionLockWatchFailedException("Cannot process events before the oldest event");
```

Comment:
Checked that this should get retried both by atlasdb-proxy and atlasdb internally.

Reply:
You might wanna add a bit more around the "this will be retried and should happen very rarely".